[doctor] metro-config deep dependency check

keith-kurak opened the pull request:
# Why
It looks like `npx expo install --check` (which is used by Doctor) will pickup a direct `metro-config` install of an invalid version, but not a deeply-installed dependency or a resolution.

Within the existing doctor test structure, this adds `metro-config` to the deep dependency checks and customizes the advice if the metro version check fails and it is contained in resolutions.

# How
I feel a little pinned by the current doctor structure in terms of making this check as nuanced as I'd like it. Namely, the versions API call is embedded inside `SupportPackageVersionCheck` along with the deep dependency version checks for all packages. Long-term, i'll move this API call (and really, all of the actual deep dependency checks, which are also expensive) outside of any checks and pass its data to all checks, who can then consider the data in constructing a very customized warning, without fetching the data again.

For now, I just added `metro-config` to the existing list of deep dependency versions to check, but also customized the warning to alert if any potential problem packages are in `resolutions`. 

# Test Plan
<img width="850" alt="image" src="https://github.com/expo/expo/assets/8053974/a36592f5-706e-485e-a52d-8ab7f02e24c4">


# Checklist

<!--
Please check the appropriate items below if they apply to your diff. This is required for changes to Expo modules.
-->

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

keith-kurak committed:
[doctor] metro-config deep dependency check

keith-kurak committed:
include specific package name in resolutions in advice

keith-kurak committed:
Update CHANGELOG.md

brentvatne commented:
let's also add:

- [metro-runtime](https://github.com/facebook/metro/tree/main/packages/metro-runtime)
- [metro](https://github.com/facebook/metro/tree/main/packages/metro-core)

this should cover most the vast majority of the problems that could arise here.

also we can change the version that we track for this to just be for `metro` and then all of these packages can be validate against that same `metro` version, since they are published at the same time as part of the metro monorepo. so any `package name that matches `/^metro.*$/` can use the same version. if there are too many moving pieces for this right now, we can add each package we check to the endpoint for this sdk

keith-kurak committed:
check metro, metro-config, metro-runtime

keith-kurak commented:
> let's also add:
> 
> * [metro-runtime](https://github.com/facebook/metro/tree/main/packages/metro-runtime)
> * [metro](https://github.com/facebook/metro/tree/main/packages/metro-core)
> 
> this should cover most the vast majority of the problems that could arise here.
> 
> also we can change the version that we track for this to just be for `metro` and then all of these packages can be validate against that same `metro` version, since they are published at the same time as part of the metro monorepo. so any `package name that matches `/^metro.*$/` can use the same version. if there are too many moving pieces for this right now, we can add each package we check to the endpoint for this sdk

Updated to just require

> let's also add:
> 
> * [metro-runtime](https://github.com/facebook/metro/tree/main/packages/metro-runtime)
> * [metro](https://github.com/facebook/metro/tree/main/packages/metro-core)
> 
> this should cover most the vast majority of the problems that could arise here.
> 
> also we can change the version that we track for this to just be for `metro` and then all of these packages can be validate against that same `metro` version, since they are published at the same time as part of the metro monorepo. so any `package name that matches `/^metro.*$/` can use the same version. if there are too many moving pieces for this right now, we can add each package we check to the endpoint for this sdk

Updated to just require `metro` in the versions endpoint, but also check `metro-config`, `metro-runtime`. We could add others, but if these are the major ones to look at, limiting how many deep checks we do helps keep things running quickly.
